src/OpenLoco/src/OpenLoco.cpp

```diff
@@ -1145,6 +1145,129 @@ namespace OpenLoco
+        {
+            return std::equal(begin(lhs), end(lhs), // will become constexpr with C++20
+                              begin(rhs),
+                              end(rhs));
+        }
+
+        template<typename T1, typename T2>
+        void bitWiseLogDivergence(const std::string entity, const T1& lhs, const T2& rhs)
+        {
+            size_t size = sizeof(T1) / sizeof(char);
+            char* array_lhs = (char*)(&lhs);
+            char* array_rhs = (char*)(&rhs);
+
+            for (size_t offset = 0; offset < size; offset++)
+            {
+                if (array_lhs[offset] != array_rhs[offset])
+                {
+                    Logging::info("DIVERGENCE");
+                    Logging::info("ENTITY: {{ {} }}", entity);
+                    Logging::info("\tOFFSET: {{ {} }}", offset);
+                    Logging::info("\t   LHS: {{ {:#x} }}", array_lhs[offset]);
+                    Logging::info("\t   RHS: {{ {:#x} }}", array_rhs[offset]);
+                }
+            }
+        }
+    }
+
+    template<typename T1, typename T2>
+    void logEntityDivergence(const std::string entity, const T1& lhs, const T2& rhs)
+    {
+        for (unsigned int offset = 0; offset < sizeof(lhs); offset++)
+            unsafe::bitWiseLogDivergence(entity + "[" + std::to_string(offset) + "]", lhs[offset], rhs[offset]);
+    }
+
+    template<typename T1, typename T2>
+    void logEntityDivergence(const std::string entity, const T1& lhs, const T2& rhs, int arraySize)
+    {
+        for (int offset = 0; offset < arraySize; offset++)
+            unsafe::bitWiseLogDivergence(entity + "[" + std::to_string(offset) + "]", lhs[offset], rhs[offset]);
+    }
+
+    template<typename T>
+    void logDivergentGameStateField(const std::string entity, int offset, const T& lhs, const T& rhs)
+    {
+        if (!unsafe::bitwise_equal(lhs, rhs))
+        {
+            Logging::info("DIVERGENCE");
+            Logging::info("ENTITY: {{ {} }}", entity);
+            Logging::info("\tOFFSET: {{ {} }}", offset);
+            Logging::info("\t   LHS: {{ {:#x} }}", lhs);
+            Logging::info("\t   RHS: {{ {:#x} }}", rhs);
+        }
+    }
+
+    void compareGameStates(const fs::path& path)
+    {
+        Logging::info("Comparing reference file {} to current GameState frame", path);
+        std::unique_ptr<S5::S5File> refFile = S5::importSave(path);
+        S5::GameState& refGameState = refFile.get()->gameState;
+
+        logDivergentGameStateField("rng_0:", 0, getGameState().rng.srand_0(), refGameState.rng[0]);
+        logDivergentGameStateField("rng_1:", 0, getGameState().rng.srand_1(), refGameState.rng[1]);
+
+        if (getGameState().flags != refGameState.flags)
+        {
+            auto flags1 = static_cast<uint32_t>(getGameState().flags);
+            auto flags2 = static_cast<uint32_t>(refGameState.flags);
+            Logging::info("DIVERGENCE");
+            Logging::info("ENTITY: {{ {} }}", "flags");
```

Comment:
I know i mentioned ENTITY as the name in the issue but that was more for a special case for entities. For non-entity things just use the name TYPE. 
For entities we would want an additional special case where it looks up the entity type and entity sub type. 
So you would have something like:
```
DIVERGENCE
TYPE: ENTITY [0] VEHICLE VEHICLE_HEAD
```
where VEHICLE is the entity type and VEHICLE_HEAD is the subtype. (Tbh you could probably skip displaying the entity type as the subtype is unique). Entity types are:
```
Vehicle,
Effect,
```
Subtypes of vehicle are:
```
VehicleHead,
Vehicle1,
Vehicle2,
Bogie,
BodyStart,
BodyContinued,
VehicleTail
```
Subtypes of Effect are:
```
        exhaust = 0, // Steam from the exhaust
        redGreenCurrency = 1,
        windowCurrency = 2,       // currency which is created in the company's colour when a transaction is made (for example the train arrives with a passengers into the station)
        vehicleCrashParticle = 3, // parts (particles) of vehicle after crash which they fall to the ground after explosion
        explosionCloud = 4,       // explosion which is created when two trains (or maybe other vehicles) crash to each other
        splash = 5,               // splash when particles after explosion land to water and creates a splash (exploding train on the bridge)
        fireball = 6,
        explosionSmoke = 7,
        smoke = 8 // Smoke from broken down train
```

Reply:
Comments most welcome.
I would like to refactor the compare code into a Compare class.